### PR TITLE
Deep Biaffine Attention Dependency Parser

### PR DESCRIPTION
**Pull request by hankcs**
## Description ##
Hi gluon-nlp team, I'd like to request for merging this SOTA dependency parser. It's an implementation of "[Deep Biaffine Attention for Neural Dependency Parsing](https://arxiv.org/pdf/1611.01734.pdf)" proposed by Dozat and Manning (2016). It's also a re-implementation of this [DyNet version](https://github.com/jcyk/Dynet-Biaffine-dependency-parser).

It replicates every detail of the original paper, somehow produces better score (maybe due to different [data pre-processing](https://github.com/hankcs/TreebankPreprocessing)). The performance of PTB-SD3.5 on a TITAN X is as follows.

```
Test: UAS 96.08% LAS 95.02% 537 sents/s
```

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

As it requires some extern corpus for testing, and I'm not familiar with your protocol. So, the test code is placed in [a main function](https://github.com/hankcs/gluon-nlp/blob/biaffine/gluonnlp/model/biaffine/parser/dep_parser.py#L288). `DepParser` is also intended for providing the common user interfaces (train, eval, parse).

You can checkout the data for debugging [there](https://github.com/hankcs/DeepBiaffineParserMXNet), or ask me for more details.

### Changes ###
- [x] Biaffine parser
- [x] Variational LSTM as in Gal and Ghahramani (2016).

## Comments ##
- The private class `_Config` might be replaced by a `Dict` object, at the loss of error checking and IO interfaces.
- The `ParserVocabulary` should be implemented with built-in `Vocab`. Since we're still discussing it, let's leave it for future work.
- My variational RNN trick is not efficient, maybe it's better to do it in C++.
- Lots of `transpose` could be reduced for efficiency. Some parts of the biaffine model could be capsuled into a [HybridBlock](https://github.com/hankcs/DeepBiaffineParserMXNet/commit/3c5e9b1168f05ce1084f0d708242a276220bd15e).
- Maybe using the built-in embeddings are preferable over my hand-crafted one.

Anyway, there are lots of work to be done, but I'd like to bring this up and discuss together with everyone who's interested.

Thanks.

### CODE REVIEW

File: gluonnlp/model/biaffine/common/data.py [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+        return len(self.words)
+
+    def __getitem__(self, index):
+        return self.words[index]
+
+    def __iter__(self):
+        return (line for line in self.words)
+
+
+class ParserVocabulary(Savable):
```
Comment (szha) [inline gluonnlp/model/biaffine/common/data.py]: Add a `@TODO` for replacing the vocabulary once #393 is addressed.

**Comment by hankcs** [comment]
Sure, I'll do them one by one. Training log and testing log are attached below.
[train.log](https://github.com/dmlc/gluon-nlp/files/2574469/train.log)
[test.log](https://github.com/dmlc/gluon-nlp/files/2574470/test.log)

I'll do the PR too.

**Comment by hankcs** [comment]
Hi @szha ,

I've moved the parser to syntactics folder then created a page for it. Anything else I can help?

**Comment by szha** [comment]
@hankcs thanks a lot! There may be a couple more style fixes that need to pass so that we can review the rendered model zoo page. I will review again once the CI passes.

**Comment by mli** [comment]
Job PR-408/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-408/6/index.html

File: scripts/syntactics/index.rst [hunk 2]
```diff
@@ -0,0 +1,78 @@
+Deep Biaffine Dependency Parser
```
Comment (szha) [inline scripts/syntactics/index.rst]: H1 should have the name of the task. Model name should occur at H2 level

File: scripts/syntactics/index.rst [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+---------------------------------
+
+This package is an implementation of `Deep Biaffine Attention for Neural Dependency Parsing <https://arxiv.org/pdf/1611.01734.pdf>`_ proposed by Dozat and Manning (2016), with SOTA accuracy.
```
Comment (szha) [inline scripts/syntactics/index.rst]: The download link needs to be added back. The zipfile name is typically the same as the folder name (i.e. `syntactics.zip`). Was there any issue when you added it?

File: scripts/syntactics/index.rst [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+Train
+^^^^^^^^^^^^^
+
+As the Penn Treebank dataset (PTB) is proprietary, we are unable to distribute it.
+If you have a legal copy, please place it in ``tests/data/biaffine/ptb``, use this `pre-processing script <https://github.com/hankcs/TreebankPreprocessing>`_ to convert it into conllx format.
+Then download `glove embeddings <http://nlp.stanford.edu/data/glove.6B.zip>`_ into ``tests/data/biaffine/embedding`` and unzip it.
```
Comment (szha) [inline scripts/syntactics/index.rst]: I understand that vocab needs to be improved before this model can adopt the existing vocab class. On the other hand, glove embedding in gluonnlp can be used independent of the vocab class, and glove embeddings are supported. Can we adapt that so that there's no need for users to download the embeddings themselves?

File: scripts/syntactics/index.rst [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+Then Run the following code to train the biaffine model.
+
+.. code-block:: python
+
+    parser = DepParser()
+    parser.train(train_file='tests/data/biaffine/ptb/train.conllx',
+                 dev_file='tests/data/biaffine/ptb/dev.conllx',
+                 test_file='tests/data/biaffine/ptb/test.conllx', save_dir='tests/data/biaffine/model',
+                 pretrained_embeddings_file='tests/data/biaffine/embedding/glove.6B.100d.txt')
```
Comment (szha) [inline scripts/syntactics/index.rst]: This is available in gluonnlp API as `gluonnlp.embedding.create('glove', 'glove.6B.100d')`

File: scripts/syntactics/index.rst (hunk 4 above)
Comment (hankcs) [reply scripts/syntactics/index.rst]: Yes, I'll use `gluonnlp.embedding.create`. But can I remove the support of custom embeddings, but only accept instances of `gluonnlp.embedding.TokenEmbedding` or their source names?

File: scripts/syntactics/index.rst (hunk 5 above)
Comment (hankcs) [reply scripts/syntactics/index.rst]: Yes, got it.

File: scripts/syntactics/index.rst (hunk 3 above)
Comment (szha) [reply scripts/syntactics/index.rst]: The download file is automatically generated during the build, so you just need to put syntactics.zip in the download. It doesn't need to be uploaded elsewhere.

File: scripts/syntactics/index.rst (hunk 4 above)
Comment (szha) [reply scripts/syntactics/index.rst]: Yes, that's perfectly fine.

**Comment by mli** [comment]
Job PR-408/9 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-408/9/index.html

**Comment by szha** [comment]
@hankcs I took a glance at the preview and the page looks great. The download link still needs to be added.

File: scripts/syntactics/index.rst [hunk 6]
```diff
@@ -0,0 +1,77 @@
+Dependency Parsing
+---------------------------------
+
```
Comment (szha) [inline scripts/syntactics/index.rst]: Add download link by adding:
```
:download:`[Download] </model_zoo/syntactics.zip>`
```

Comment (hankcs) [reply scripts/syntactics/index.rst]: Now link has been added, but NMT link broke, failed the test.

```
Warning, treated as error:

/var/lib/jenkins/workspace/gluon-nlp-docs/docs/model_zoo/machine_translation/index.rst.rst:43:broken link: https://apache-mxnet.s3.cn-north-1.amazonaws.com.cn/gluon/models/transformer_en_de_512_WMT2014-14bd361b.zip (('Connection broken: OSError("(104, \'ECONNRESET\')",)', OSError("(104, 'ECONNRESET')",)))

Makefile:171: recipe for target 'linkcheck' failed
```

It must be occasionally, I can download it using browser. What shall I do?

Comment (szha) [reply scripts/syntactics/index.rst]: @hankcs thanks. Sorry about it as we just recently added link check and are still ironing out these intermittent issues. For now, you can delete the last line of https://github.com/dmlc/gluon-nlp/blob/master/scripts/machine_translation/index.rst. I will address it in a separate PR.

Comment (szha) [reply scripts/syntactics/index.rst]: Or, you can wait for https://github.com/dmlc/gluon-nlp/pull/414 in which the linkcheck is not enforced on PRs.

Comment (hankcs) [reply scripts/syntactics/index.rst]: > @hankcs thanks. Sorry about it as we just recently added link check and are still ironing out these intermittent issues. For now, you can delete the last line of https://github.com/dmlc/gluon-nlp/blob/master/scripts/machine_translation/index.rst. I will address it in a separate PR.

I've temporarily delete that link, thanks.

**Comment by mli** [comment]
Job PR-408/11 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-408/11/index.html

File: scripts/syntactics/common/utils.py [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+import mxnet.ndarray as nd
+import numpy as np
+from mxnet.gluon import rnn
+from mxnet.gluon.contrib.rnn import VariationalDropoutCell
+
+from .data import ParserVocabulary
+from .tarjan import Tarjan
+
+
+class Progbar(object):
```
Comment (sxjscience) [inline scripts/syntactics/common/utils.py]: Do we have to copy this function?

File: scripts/syntactics/common/utils.py [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+    gpu = int(os.environ.get('MXNET_GPU', default=0))
+    return mx.gpu(gpu)
+
+
+def init_logger(root_dir, name="train.log"):
+    """
+    Initialize a logger
+    :param root_dir: directory for saving log
+    :param name: logger name
+    :return: a logger
```
Comment (sxjscience) [inline scripts/syntactics/common/utils.py]: Use numpy docstyle

File: scripts/syntactics/common/utils.py [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+
+
+def reshape_fortran(tensor, shape):
+    """
+    The missing Fortran reshape for mx.NDArray
+    :param tensor: source tensor
+    :param shape: desired shape
+    :return: reordered result
+    """
+    return tensor.T.reshape(tuple(reversed(shape))).T
```
Comment (sxjscience) [inline scripts/syntactics/common/utils.py]: We have adopted too many functions from https://github.com/tdozat/Parser .

File: scripts/syntactics/common/utils.py [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+    y = y.transpose([2, 1, 0])  # May cause performance issues
+    lin = lin.transpose([2, 1, 0])
+    blin = nd.batch_dot(lin, y, transpose_b=True)
+    blin = blin.transpose([2, 1, 0])
+    if num_outputs > 1:
+        blin = reshape_fortran(blin, (seq_len, num_outputs, seq_len, batch_size))
+    return blin
+
+
+def orthonormal_initializer(output_size, input_size, debug=False):
```
Comment (sxjscience) [inline scripts/syntactics/common/utils.py]: I'm not sure if it's the same as the `Orthogonal` initializer in MXNet. https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/initializer.py#L521

File: scripts/syntactics/parser/biaffine_parser.py [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+        rel_targets : mxnet.ndarray.NDArray
+            gold rel indices of seq_len x batch_size
+        Returns
+        -------
+        tuple
+            (arc_accuracy, rel_accuracy, overall_accuracy, loss) when training, else if given gold target
+        then return arc_accuracy, rel_accuracy, overall_accuracy, outputs, otherwise return outputs, where outputs is a
+        list of (arcs, rels).
+        """
+        is_train = autograd.is_training()
```
Comment (sxjscience) [inline scripts/syntactics/parser/biaffine_parser.py]: I suggest moving this logic to the main training/inference loops.

File: scripts/syntactics/parser/biaffine_parser.py [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+        if is_train:
+            wm, tm = self.generate_emb_mask(seq_len, batch_size)
+            emb_inputs = nd.concat(nd.multiply(wm, word_embs), nd.multiply(tm, tag_embs), dim=2)
+        else:
+            emb_inputs = nd.concat(word_embs, tag_embs, dim=2)  # seq_len x batch_size
+
+        top_recur = biLSTM(self.f_lstm, self.b_lstm, emb_inputs, batch_size,
+                           dropout_x=self.dropout_lstm_input if is_train else 0)
+        if is_train:
+            top_recur = nd.Dropout(data=top_recur, axes=[0], p=self.dropout_mlp)
```
Comment (sxjscience) [inline scripts/syntactics/parser/biaffine_parser.py]: Dropout will be automatically turned off in the testing mode so there's no need to add this `if`.

File: scripts/syntactics/common/utils.py [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+            rel_preds[roots] = new_rel_preds
+            rel_preds[new_root] = root
+        return rel_preds
+    else:
+        rel_probs[:, ParserVocabulary.PAD] = 0
+        rel_preds = np.argmax(rel_probs, axis=1)
+        return rel_preds
+
+
+def reshape_fortran(tensor, shape):
```
Comment (sxjscience) [inline scripts/syntactics/common/utils.py]: Why do we need this?

File: scripts/syntactics/parser/biaffine_parser.py [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
+            rel_preds = partial_rel_logits.argmax(0)
+            targets_1D = flatten_numpy(rel_targets)
+            rel_correct = np.equal(rel_preds.asnumpy(), targets_1D).astype(np.float32) * mask_1D
+            rel_accuracy = np.sum(rel_correct) / num_tokens
+            losses = self.softmax_loss(partial_rel_logits, nd.array(targets_1D))
+            rel_loss = nd.sum(losses * mask_1D_tensor) / num_tokens
+
+        if not is_train:
+            rel_probs = np.transpose(np.reshape(nd.softmax(flat_rel_logits.transpose([1, 0, 2]), axis=0).asnumpy(),
+                                                (self._vocab.rel_size, seq_len, seq_len, batch_size), 'F'))
```
Comment (sxjscience) [inline scripts/syntactics/parser/biaffine_parser.py]: We can move the `if` logics above into the training/inference loops.

File: scripts/syntactics/common/utils.py (hunk 7 above)
Comment (hankcs) [reply scripts/syntactics/common/utils.py]: Not necessary, this bar is intuitive, this is the only reason. Shall I remove that?

File: scripts/syntactics/common/utils.py (hunk 8 above)
Comment (hankcs) [reply scripts/syntactics/common/utils.py]: I forgot this one, now it's done.

File: scripts/syntactics/common/utils.py (hunk 9 above)
Comment (hankcs) [reply scripts/syntactics/common/utils.py]: Probably true. Most of them are kept to ensure the consistency with original implementation. I only focused on the DL part. Shall I remove those unnecessary KMeans bucketing and Tarjan stuff? As the parser algorithm is good enough, the accuracy won't drop much.

File: scripts/syntactics/common/utils.py (hunk 10 above)
Comment (hankcs) [reply scripts/syntactics/common/utils.py]: Not the same, not mentioned in parser paper. I'll replace this with the SVD implementation, such that a black box is removed.

**Comment by mli** [comment]
Job PR-408/13 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-408/13/index.html

File: scripts/syntactics/parser/biaffine_parser.py (hunk 11 above)
Comment (hankcs) [reply scripts/syntactics/parser/biaffine_parser.py]: Yes, done.

**Comment by mli** [comment]
Job PR-408/14 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-408/14/index.html

**Comment by mli** [comment]
Job PR-408/15 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-408/15/index.html

**Comment by mli** [comment]
Job PR-408/16 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-408/16/index.html
